src/asynchronous/stream.rs

```diff
@@ -1,137 +1,482 @@
+{
+    pub fn new(inner: StreamInner) -> Self {
+        Self {
+            inner: inner.split().0,
+            _send: PhantomData,
+        }
+    }
+
+    pub async fn send(&self, resp: &P) -> Result<()> {
+        let msg_buf = resp
+            .encode()
+            .map_err(err_to_others_err!(e, "Encode message failed."))?;
+        self.inner.send(msg_buf).await
+    }
+}
+
+pub struct ClientStreamReceiver<P> {
+    inner: StreamReceiver,
+    _recv: PhantomData<P>,
+}
+
+impl<P> ClientStreamReceiver<P>
+where
+    P: Codec,
+    <P as Codec>::E: std::fmt::Display,
+{
+    pub fn new(inner: StreamInner) -> Self {
+        Self {
+            inner: inner.split().1,
+            _recv: PhantomData,
+        }
+    }
+
+    pub async fn recv(&mut self) -> Result<Option<P>> {
+        let res = self.inner.recv().await;
+        if matches!(res, Err(Error::EOF)) {
+            return Ok(None);
+        }
+        let msg_buf = res?;
+        P::decode(&msg_buf)
+            .map_err(err_to_others_err!(e, "Decode message failed."))
+            .map(Some)
+    }
+}
+
+pub struct ServerStreamReceiver<Q> {
+    inner: StreamReceiver,
+    _recv: PhantomData<Q>,
+}
+
+impl<Q> ServerStreamReceiver<Q>
+where
+    Q: Codec,
+    <Q as Codec>::E: std::fmt::Display,
+{
+    pub fn new(inner: StreamInner) -> Self {
+        Self {
+            inner: inner.split().1,
+            _recv: PhantomData,
+        }
+    }
+
+    pub async fn recv(&mut self) -> Result<Option<Q>> {
+        let res = self.inner.recv().await;
+        if matches!(res, Err(Error::EOF)) {
+            return Ok(None);
+        }
+        let msg_buf = res?;
+        Q::decode(&msg_buf)
+            .map_err(err_to_others_err!(e, "Decode message failed."))
+            .map(Some)
+    }
+}
 
-    tx.send(buf)
+async fn _recv(rx: &mut ResultReceiver) -> Result<GenMessage> {
+    rx.recv()
         .await
-        .map_err(err_to_others_err!(e, "Send packet to sender error "))
+        .unwrap_or_else(|| Err(Error::Others("Recive packet from recver error".to_string())))
 }
 
-pub(crate) async fn respond_with_status(
-    tx: tokio::sync::mpsc::Sender<Vec<u8>>,
-    stream_id: u32,
-    status: Status,
-) -> Result<()> {
-    let mut res = Response::new();
-    res.set_status(status);
-    let mut body = get_response_body(&res)?;
-
-    let mh = MessageHeader {
-        length: body.len() as u32,
-        stream_id,
-        type_: MESSAGE_TYPE_RESPONSE,
-        flags: 0,
-    };
-
-    let mut buf = header_to_buf(mh);
-    buf.append(&mut body);
-
-    tx.send(buf)
+async fn _send(tx: &MessageSender, msg: GenMessage) -> Result<()> {
+    tx.send(msg)
         .await
-        .map_err(err_to_others_err!(e, "Send packet to sender error "))
+        .map_err(|e| Error::Others(format!("Send data packet to sender error {:?}", e)))
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Kind {
+    Client,
+    Server,
+}
+
+#[derive(Debug)]
+pub struct StreamInner {
```

Comment:
It's weird something called xxxInner be public

Reply:
This type supports all send message and receive message operations. It is used in the server-side `trait StreamHandler` and in the client-side `new_stream`. 
then It is wrapped in the generated code as the type that will eventually be used by the user. 

I'm not good at making a new name. Any suggestion from your side?

Reply:
StreamCore ? StreamCommon? uh... StreamInner is not so bad now I think